crates/storage/src/address.rs

```diff
@@ -0,0 +1,480 @@
+    /// ```
+    ///
+    #[must_use]
+    pub fn is_root(&self) -> bool {
+        self.depth() == 0
+    }
+
+    /// Joins two [`Path`]s.
+    ///
+    /// Joins the [`Path`] with another [`Path`], returning a new [`Path`] that
+    /// is the concatenation of the two.
+    ///
+    /// # Parameters
+    ///
+    /// * `other` - The other [`Path`] to join with.
+    ///
+    /// # Errors
+    ///
+    /// An error will be returned if the resulting path would be too long.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path1 = Path::new("::root::node").unwrap();
+    /// let path2 = Path::new("::leaf").unwrap();
+    /// let joined = path1.join(&path2).unwrap();
+    /// assert_eq!(joined.to_string(), "::root::node::leaf");
+    /// ```
+    ///
+    pub fn join(&self, other: &Self) -> Result<Self, PathError> {
+        if self.path.len().saturating_add(other.path.len()) > 255 {
+            return Err(PathError::Overflow);
+        }
+        let mut path: Flexstr<255> = Flexstr::new();
+        let _: bool = path.push_str(&self.path);
+        let _: bool = path.push_str(&other.path);
+        let mut offsets = self.offsets.clone();
+        #[allow(clippy::cast_possible_truncation)] // Can't occur here
+        offsets.push(self.path.len() as u8);
+        #[allow(clippy::cast_possible_truncation)] // Can't occur here
+        offsets.extend(
+            other
+                .offsets
+                .iter()
+                .map(|&offset| offset.saturating_add(self.path.len() as u8)),
+        );
+        Ok(Self { offsets, path })
+    }
+
+    /// The last segment of the [`Path`].
+    ///
+    /// Returns the last segment of the path, which is the bottom-most in the
+    /// hierarchy expressed by the path.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.last(), "leaf");
+    /// ```
+    ///
+    #[must_use]
+    pub fn last(&self) -> &str {
+        if self.offsets.is_empty() {
+            &self.path
+        } else {
+            self.offsets
+                .last()
+                .map_or(&self.path, |&offset| &self.path[offset as usize..])
+        }
+    }
+
+    /// The parent of the [`Path`].
+    ///
+    /// Returns the parent of the [`Path`], which is the path with the last
+    /// segment removed. If the path is already at the root, then `None` is
+    /// returned.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use calimero_storage::address::Path;
+    /// let path = Path::new("::root::node::leaf").unwrap();
+    /// assert_eq!(path.parent().unwrap().to_string(), "::root::node");
+    /// assert_eq!(Path::new("::root").unwrap().parent(), None);
+    /// ```
+    ///
+    #[must_use]
+    pub fn parent(&self) -> Option<Self> {
+        let mut clone = self.clone();
```

Comment:
Cloning all of `Self` is typically not the right thing to do when you don't need all of it as it will reallocate all of `Self`'s heap-backed capacity, which in certain cases won't be necessary.

Like going from `::a::b::c::d::e::f` (which would've allocated a `Vec` capable of holding 8 offsets) to `::a::b::c::d::e`, (whose offsets would fit comfortably in one capable of holding 4), you'll still have a `Vec` with capacity for 8.

It's fairly trivial, but from my understanding of how this module is written, it seems like optimization is a key piece, in which case, I figured I'd mention.

Reply:
That's insightful. I don't think the `clone()` is all that bad, as it needs to duplicate all but the last segment - but, it is technically unnecessary as we have the offsets and so could use them for the calculation 🤔